fendermint/app/src/cmd/run.rs

```diff
@@ -367,3 +418,60 @@ fn to_address(sk: &SecretKey, kind: &AccountKind) -> anyhow::Result<Address> {
         AccountKind::Ethereum => Ok(Address::from(EthAddress::new_secp256k1(&pk)?)),
     }
 }
+
+async fn dispatch_resolver_events(
+    mut rx: tokio::sync::broadcast::Receiver<ResolverEvent<AppVote>>,
+    parent_finality_votes: VoteTally,
+    topdown_enabled: bool,
+) {
+    loop {
+        match rx.recv().await {
+            Ok(event) => match event {
+                ResolverEvent::ReceivedPreemptive(_, _) => {}
+                ResolverEvent::ReceivedVote(vote) => {
+                    dispatch_vote(*vote, &parent_finality_votes, topdown_enabled).await;
+                }
+            },
+            Err(RecvError::Lagged(n)) => {
+                tracing::warn!("the resolver service skipped {n} gossip events")
+            }
+            Err(RecvError::Closed) => {
+                tracing::error!("the resolver service stopped receiving gossip");
+                return;
+            }
+        }
+    }
+}
+
+async fn dispatch_vote(
+    vote: VoteRecord<AppVote>,
+    parent_finality_votes: &VoteTally,
+    topdown_enabled: bool,
+) {
+    match vote.content {
+        AppVote::ParentFinality(f) => {
+            if !topdown_enabled {
+                tracing::debug!("ignoring vote; topdown disabled");
+                return;
+            }
+            let res = atomically_or_err(|| {
+                parent_finality_votes.add_vote(
+                    vote.public_key.clone(),
+                    f.height,
+                    f.block_hash.clone(),
+                )
+            })
+            .await;
+
+            match res {
+                Ok(_) => {}
+                Err(e @ VoteError::Equivocation(_, _, _, _)) => {
+                    tracing::warn!(error = e.to_string(), "failed to handle vote");
+                }
+                Err(e) => {
+                    tracing::debug!(error = e.to_string(), "failed to handle vote")
```

Comment:
Why is this debug? Are general errors normal here?

Reply:
Yeah this is a grey area, I should add a comment. I didn't want to handle all possible errors in the enum to keep the noise down here, but maybe that would save some head cratching as well as force us to pick levels for any future errors:
* `Unintitialized` we can safely ignore, there will be other votes once the system is up and running; unless it keeps repeating and we're wondering what's happening, in which case during debug it is helpful to realise that initialisation has failed
* `UnpoweredValidator` can happen if we get a vote from someone who isn't a validator according to our view, but they could have been one in a previous block, or they might be one in the next block; or they might truly just publish trolling votes. Unless there is a way to identify vandals, I wouldn't give them the power to spam the logs, but in debug it can help to see this is happening.
* `UnexpectedBlock` will not be returned by this method.